server/coordinator.go

```diff
@@ -446,3 +455,85 @@ func collectOperatorCounterMetrics(op Operator) {
 		operatorCounter.WithLabelValues(op.GetName(), op.GetState().String()).Add(1)
 	}
 }
+
+type heartbeatStream interface {
+	Send(*pdpb.RegionHeartbeatResponse) error
+}
+
+type streamUpdate struct {
+	storeID uint64
+	stream  heartbeatStream
+}
+
+type heartbeatStreams struct {
+	ctx       context.Context
+	clusterID uint64
+	streams   map[uint64]heartbeatStream
+	msgCh     chan *pdpb.RegionHeartbeatResponse
+	streamCh  chan streamUpdate
+}
+
+func newHeartbeatStreams(ctx context.Context, clusterID uint64) *heartbeatStreams {
+	localCtx, _ := context.WithCancel(ctx)
+	hs := &heartbeatStreams{
+		ctx:       localCtx,
+		clusterID: clusterID,
+		streams:   make(map[uint64]heartbeatStream),
+		msgCh:     make(chan *pdpb.RegionHeartbeatResponse, regionheartbeatSendChanCap),
+		streamCh:  make(chan streamUpdate, 1),
+	}
+	go hs.run()
+	return hs
+}
+
+func (s *heartbeatStreams) run() {
+	for {
+		select {
+		case update := <-s.streamCh:
+			s.streams[update.storeID] = update.stream
+		case msg := <-s.msgCh:
+			storeID := msg.GetTargetPeer().GetStoreId()
+			if stream, ok := s.streams[storeID]; ok {
+				if err := stream.Send(msg); err != nil {
+					log.Errorf("send heartbeat message fail: %v", err)
+					delete(s.streams, storeID)
+					regionHeartbeatCounter.WithLabelValues("down", "err")
+				} else {
+					regionHeartbeatCounter.WithLabelValues("down", "ok")
```

Comment:
_push_ for server -> client, and what for client -> server? @huachaohuang @siddontang